### ISSUE TITLE
[Bug]: Regression introduced in 7.15.3

### ISSUE THREAD

**Issue by jhg023** [react-hook-form/react-hook-form#6659]
### Version Number

7.15.3

### Codesandbox/Expo snack

I'm currently unable to reproduce issue in Codesandbox, but I'm going to spend more time trying to reproduce it in the morning. I explain the issue in further detail below.

### Steps to reproduce

![image](https://user-images.githubusercontent.com/18515648/135049113-7b307e29-53f3-4f60-89e1-5bc59cfa29ee.png)
The above change was not backwards compatible, and seemed to have broken the following code of mine:
```
reset(
  {
    filters: filters,
  },
  { keepValues: true }
);
```

# Steps
1. Make change to form element (i.e. type something in text `<input>`, for example).
2. `isDirty` is set to `true`, triggering a `useEffect`.
3. Invoke the above `reset` snippet within the aforementioned `useEffect`.
4. The form element that was modified should still be focused (blinking cursor should not disappear).
4. Make another change to the same form element. Any further changes **do not** set `isDirty` to `true`, which is not the correct behavior.

### Expected behaviour

I expect `isDirty` to be set to `true` on any change to the form element, regardless of whether or not the form is `reset` with `{ keepValues: true }`.

I cannot remove `{ keepValues: true }`, because then the form element loses focus when `reset` is called.

I had to downgrade to version `7.15.2` for it to work as expected, so version `7.15.3` was where the breaking change was made.

In the image that I posted above, you can see that `_formValues` is not set if `keepStateOptions.keepValues` is set to `true`, which is different than the behavior in version `7.15.2`.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
Related fix: 🐞 fix #6482 reset formValues with shouldUnregister: true (#6512)

**Comment by bluebill1049** [comment]
is this still an issue?

**Comment by jhg023** [comment]
> is this still an issue?

I don't use `shouldUnregister: true` on my form, but yes, it's still an issue even on the latest version (`7.16.1`) of `react-hook-form`.

The only solution that I've found so far has been to downgrade back to version `7.15.2`.

I'll attempt to reproduce the issue in a Codesandbox now.

**Comment by bluebill1049** [comment]
sure, let me know when got a reproducible CSB. 👍

**Comment by jhg023** [comment]
I'm not having much luck reproducing the issue on a CodeSandbox.

However, I've discovered that after calling `reset(..., { keepValues: true });`, my component does not render whenever any subsequent change is made to the form's text field.

By downgrading the `7.15.2`, the component renders just fine after invoking `reset(..., { keepValues: true });` (by continuing to type in the form's text field).

Another solution to my issue would be to reset the form without losing focus on the form element (and without specifying `{ keepValues: true }`).

Does there currently exist any mechanism to reset the form without losing focus on the form element?

If not, I think I'll just have to scrap the form entirely on this component and handle the states (and `onChange` callbacks) manually.

**Comment by bluebill1049** [comment]
> Does there currently exist any mechanism to reset the form without losing focus on the form element?

we are running the native reset API, if you want to just update an input or inputs, perhaps you want to use `setValue` instead?

**Comment by jhg023** [comment]
> we are running the native reset API, if you want to just update an input or inputs, perhaps you want to use `setValue` instead?

Thanks for the quick response, and for bearing with me for this issue.

I had tried `setValue`, but it seems to have the opposite effect (the form element loses focus after calling `setValue`, and `isDirty` is not set back to `false` like it is with `reset` on `7.15.2`).

I decided I'm just going to reimplement this component without a form, as the component itself doesn't actually have any explicit `submit` functionality where a form would be useful.

I'll happily blame myself as the cause of this issue, as I'm not able to reproduce it in a CodeSandbox. Also, I'm still using `react-hook-form` all throughout my project, so this issue won't dissuade me from continuing to use this awesome library :)

Thanks again for your help, and I really appreciate the effort you put into this project!
